packages/ipfs-http-client/src/lib/core.js

```diff
@@ -1,58 +1,59 @@
 'use strict'
 /* eslint-env browser */
 const Multiaddr = require('multiaddr')
-const toUri = require('multiaddr-to-uri')
 const { isBrowser, isWebWorker } = require('ipfs-utils/src/env')
-const { URL } = require('iso-url')
 const parseDuration = require('parse-duration').default
 const log = require('debug')('ipfs-http-client:lib:error-handler')
 const HTTP = require('ipfs-utils/src/http')
 const merge = require('merge-options')
+const toUrlString = require('ipfs-core-utils/src/to-url-string')
 
 /**
- * @param {any} input
- * @returns {input is Multiaddr}
- */
-const isMultiaddr = (input) => {
-  try {
-    Multiaddr(input) // eslint-disable-line no-new
-    return true
-  } catch (e) {
-    return false
-  }
-}
-
-/**
- * @param {any} options
+ * @param {ClientOptions|URL|Multiaddr|string} options
  * @returns {ClientOptions}
  */
-const normalizeInput = (options = {}) => {
-  if (isMultiaddr(options)) {
-    options = { url: toUri(options) }
-  } else if (typeof options === 'string') {
-    options = { url: options }
-  }
+const normalizeOptions = (options = {}) => {
+  let url
+  let opts
+
+  if (typeof options === 'string' || Multiaddr.isMultiaddr(options)) {
+    url = new URL(toUrlString(options))
+    opts = {}
+  } else if (options instanceof URL) {
+    url = options
+    opts = {}
+  } else if (typeof options.url === 'string' || Multiaddr.isMultiaddr(options.url)) {
+    url = new URL(toUrlString(options.url))
+    opts = options
+  } else if (options.url instanceof URL) {
+    url = options.url
+    opts = options
+  } else {
+    opts = options || {}
 
-  const url = new URL(options.url)
-  if (options.apiPath) {
-    url.pathname = options.apiPath
-  } else if (url.pathname === '/' || url.pathname === undefined) {
-    url.pathname = 'api/v0'
-  }
-  if (!options.url) {
     if (isBrowser || isWebWorker) {
-      url.protocol = options.protocol || location.protocol
-      url.hostname = options.host || location.hostname
-      url.port = options.port || location.port
+      const protocol = (opts.protocol || location.protocol).replace(':', '')
+
+      url = new URL(`${protocol}://${opts.host || location.hostname}:${opts.port || location.port}`)
     } else {
-      url.hostname = options.host || 'localhost'
-      url.port = options.port || '5001'
-      url.protocol = options.protocol || 'http'
+      url = new URL(`${opts.protocol || 'http'}://${opts.host || 'localhost'}:${opts.port || '5001'}`)
     }
   }
-  options.url = url
 
-  return options
+  if (opts.apiPath) {
+    url.pathname = opts.apiPath
+  } else if (url.pathname === '/' || url.pathname === undefined) {
+    url.pathname = 'api/v0'
+  }
+
+  return {
+    ...opts,
+    host: url.host,
+    protocol: url.protocol.replace(':', ''),
```

Comment:
Do we need this with the previous replace?

Reply:
Yes because URL objects add a `:` on the end of `.protocol`:

```js
new URL('http://example.com').protocol // 'http:'

new URL('http:://example.com') // throws

const u = new URL('http://example.com')
u.protocol = 'https::::::'
u.protocol // 'https:'

u.protocol = 'http'
u.protocol // 'http:'
```